### ISSUE TITLE
[API] override dependency by passing a callback

### ISSUE THREAD

**Issue by felipesanches** [googlefonts/fontbakery#3179]
After a videocall with @RosaWagner this morning, we've decided that we want to run the **designer_profiles** check on the entire collection as a github action on the google/fonts repo.

This check queries the google/fonts repo by making HTTP requests to the static files hosted on github. But since this specific use will be running on the actual repo, all the designers catalog files are available locally (in the same repo) so there's no need to waste time making HTTP requests.

So this means that we could customize the check behaviour by replacing the routine that fetches the profile info. I propose that we may be able to do so by invoking the check and passing an over-ride to the condition (as we're already able to do, and we extensively use this feature in our code-tests) but passing a python funciton instead of a fixed value. Such function would then be used to compute the condition, enabling users of the API to customize them.

This is a draft of what such code would look like:

```python
from fontbakery.profiles import googlefonts as googlefonts_profile
from fontbakery.api import FontBakeryCheck 
check = FontBakeryCheck(googlefonts_profile,
                                           "com.google.fonts/check/metadata/designer_profiles")
def fetch_designer_info(designer_name):
    # here we load and parse the info.pb file
    # for this designer directly from the
    # local repo and return a ProtoBuf object.
    return info

ttFont = TTFont("some-fontfile.ttf"))
check(ttFont, {designer_info: fetch_designer_info})
```

**Note:**
`from fontbakery.api import FontBakeryCheck`
is pseudo-code equivalent to this in our current code-base:
`from fontbakery.codetesting import CheckTester as FontBakeryCheck`

**Comment by felipesanches** [comment]
naturaly, the original check implementation would need to be refactored so that the info.pb file is read via a condition called `designer_info` in order to enable this kind of customization.

**Comment by felipesanches** [comment]
The invoked callback would have to be given access to the same context of the original check & condition so that it can manipulate `ttFont`, or any other variables involved in the checking routine.

**Comment by felipesanches** [comment]
While this issue illustrates the feature by describing a solution to a specific need, it actually proposes a generic feature that can be freely used on the fontbakery API for python scripting.

**Comment by graphicore** [comment]
> the fontbakery API for python scripting

What is that and who want's to use it how? I mean, in a way, all of fontbakery is an API for Python since it's written in Python.

regarding the implementation:

Within `CheckTester`, if there are overrides for `@conditions`, I suggest creating a copy/clone of the original profile instance, to avoid unwanted side effects. There's (yet) no `newProfile = profile.clone()` but

```py
newProfile = Profile()
newProfle.merge_profile(profile)
```

 should do the job, then, all namespace items can be overridden e.g. with `newProfle.add_to_namespace(...)` or `newProfle.register_condition()` when the condition has its `force` attribute set to `True`.


> The invoked callback would have to be given access to the same context of the original check & condition so that it can manipulate ttFont, or any other variables involved in the checking routine.

Hmm, nothing should manipulate the `ttFont` instances, but the `@condition ttFont` could be overridden as described above.

**Comment by felipesanches** [comment]
> Hmm, nothing should manipulate the ttFont instances

I said "manipulate" meaning read-only access.

**Comment by felipesanches** [comment]
>  in a way, all of fontbakery is an API for Python since it's written in Python.

yes indeed. What I meant here was an API intentionally designed to be used for performing font checks from within third-party python scripts, instead of running checks by invoking the `fontbakery` main program.

> and who want's to use it how?

One use case would be to run one check (such as the **designer_profiles** one we've been talking about) across an entire collection and reporting the results in aggregate. For instance, we could make a helper script that outputs a list of names of all designers who lack an entry on the Google Fonts designers catalog in the format seen [in this comment](https://github.com/googlefonts/fontbakery/pull/3091#issuecomment-730015421).

**Comment by felipesanches** [comment]
the lists that I posted in that issue were generated manually by running fontbakery and then filtering the markdown output with a few unix shell tools like grep, sort, uniq, etcetera... A helper script will make life easier for this type of customized collection-wide reporting.

**Comment by graphicore** [comment]
I'm not sure how I would implement that, maybe with a minimalist terminal reporter, that uses the check runner protocol directly, but only prints the desired information.

**Comment by felipesanches** [comment]
I'm drafting such a script right now. I will post it here later. But it looks really similar to the code I wrote at the description above: https://github.com/googlefonts/fontbakery/issues/3179#issue-815755956 with aggregate reporting implemented in that `do_something` function.
